contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,517 @@
+        } else {
+            self._index(pos)._value.get()
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or `U160::ZERO` if there is none.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to lookup.
+    pub fn upper_lookup(&self, key: U96) -> U160 {
+        let len = self.length();
+        let pos = self._upper_binary_lookup(key, U256::ZERO, len);
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._index(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or `U160::ZERO` if there is none.
+    ///
+    /// This is a variant of [`Self::upper_lookup`] that is optimized to find
+    /// "recent" checkpoint (checkpoints with high keys).
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to query.
+    pub fn upper_lookup_recent(&self, key: U96) -> U160 {
+        let len = self.length();
+
+        let mut low = U256::ZERO;
+        let mut high = len;
+
+        if len > uint!(5_U256) {
+            let mid = len - len.sqrt();
+            if key < self._index(mid)._key.get() {
+                high = mid;
+            } else {
+                low = mid + uint!(1_U256);
+            }
+        }
+
+        let pos = self._upper_binary_lookup(key, low, high);
+
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._index(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the most recent checkpoint, or `U160::ZERO` if
+    /// there are no checkpoints.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest(&self) -> U160 {
+        let pos = self.length();
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._index(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns whether there is a checkpoint in the structure (i.g. it
+    /// is not empty), and if so, the key and value in the most recent
+    /// checkpoint.
+    /// Otherwise, [`None`] will be returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest_checkpoint(&self) -> Option<(U96, U160)> {
+        let pos = self.length();
+        if pos == U256::ZERO {
+            None
+        } else {
+            let checkpoint = self._index(pos - uint!(1_U256));
+            Some((checkpoint._key.get(), checkpoint._value.get()))
+        }
+    }
+
+    /// Returns the number of checkpoints.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn length(&self) -> U256 {
```

Comment:
nit: Should we name this `len` to be consistent with Rust?

Reply:
thnx! changed it